docs/how-to/create_plugin.rst

```diff
@@ -0,0 +1,131 @@
+~~~~~~~
+First, we'll have to pull in our imports. Since this is such a simple plugin,
+the list of necessary imports is quite small:
+
+.. code:: py
+
+    from typing import Literal
+    from overrides import override
+
+    from .base import Plugin
+    from .properties import PluginProperties
+
+Defining a properties class
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+All plugins make use of a "properties" class to describe their unique parameters.
+The dump class is very basic, so we only need to define the required parameters:
+
+.. code:: py
+
+    class NewDumpPluginProperties(PluginProperties, frozen=True):
+        """The part properties used by the new dump plugin."""
+
+        # The human-friendly name of the plugin
+        plugin: Literal["new-dump"] = "new-dump"
+        # The "source" parameter is special - we'll talk about it more soon.
+        source: str
+
+Creating the plugin class
+~~~~~~~~~~~~~~~~~~~~~~~~~
+The plugin class itself contains the functionality of the plugin. Start by
+declaring the properties class:
+
+.. code:: py
+
+    class NewDumpPlugin(Plugin):
+        """Copy the content from the part source."""
+        
+        properties_class = NewDumpPluginProperties
+    
+Next, we'll have to start defining the special methods that make this plugin
+"work". Since this is such a simple plugin, many of these methods are empty:
+
+.. code:: py
+
+    @override
+    def get_build_snaps(self) -> set[str]:
+        """Return a set of required snaps to install in the build
+        environment."""
+        return set()
+
+    @override
+    def get_pull_commands(self) -> list[str]:
+        """Return a list of commands to retrieve dependencies during the
+        pull step."""
+        return []
+
+    @override
+    def get_build_packages(self) -> set[str]:
+        """Return a set of required packages to install in the build
+        environment."""
+        return set()
+
+    @override
+    def get_build_environment(self) -> dict[str, str]:
+        """Return a dictionary with the environment to use in the build
+        step."""
+        return {}
+
+The last method that we have to define, however, is where the actual build
+commands are defined. These are the exact commands executed by a subprocess
+during the build process:
+
+.. code:: py
+
+    @override
+    def get_build_commands(self) -> list[str]:
+        """Return a list of commands to run during the build step."""
+        install_dir = self._part_info.part_install_dir
+        return [f'cp --archive --link --no-dereference . "{install_dir}"']
+
+Notice we don't require a user-provided source directory/file - this is because
+we let the ``source`` parameter do the heavy-lifting. The source parameter is a
+mandatory parameter that specifies the working files of a part. It has a lot of
+specialised logic for different file types ranging from tar files to git
+repositories, so we can leverage that to get much stronger functionality out of
+our otherwise simple plugin.
+
```

Comment:
Can we provide a few more details about other bits the developer might want to take action on, like the first steps with `get_build_snaps`? My concern is that with all this placeholder code there's not much for our audience to go on.

Reply:
I left the doc comments in those functions to hopefully cover that problem. I'm having some difficulty balancing "only explaining what we need" and "covering all the bases" without creating too complex of an example. I can give a small definitions list blurb of the functions, perhaps? Taking suggestions here!

Reply:
There are a few things we could do:

1. Chunk this section into steps. Each new paragraph could be its own step, so I count three. This advice applies to all these sections, BTW.
2. After "many of these methods are empty", we could _direct_ the reader to "Fill in the behaviour of these methods", for example.
3. Expanding on # 2, we could then bring back your earlier Java example, and suggest that for that hypothetical plugin,  `get_build_packages` would need to be extended to shepherd in the installation of Java runtimes and dependencies for the plugin.

If it helps, here's a [recent example](https://canonical-snapcraft.readthedocs-hosted.com/en/latest/howto/craft-for-platforms/example-gtk4-app/) of how I described writing code (well, configurations) as steps.